### PR TITLE
Expose CompositeDrawable's subtree specific positional input conditional

### PR DESCRIPTION
**Pull request by nyquillerium**
Previously, having `Masking` disabled on a `CompositeDrawable` means children contained inside it are able to receive positional input even if its from outside of the `CompositeDrawable`'s `DrawRectangle`. 

This behavior is undesirable for some scenarios, such as if we want to use a `Container` to denote the area for where input should be allowed, such as in https://github.com/ppy/osu/issues/5452.

This exposes a new virtual property that can be overridden to denote any `CompositeDrawable` as a positional input-confining drawable, forcing a `ReceivePositionalInputAt` check, as well as makes `ScrollContainers` override it by default.

### CODE REVIEW

File: osu.Framework/Graphics/Containers/CompositeDrawable.cs [hunk 1]
```diff
@@ -1221,12 +1221,18 @@ internal override bool BuildNonPositionalInputQueue(List<Drawable> queue, bool a
             return true;
         }
 
+        /// <summary>
+        /// Whether or not this Drawable should block its children from receiving input outside of its
+        /// draw rectangle regardless of whether or not they are masked away.
+        /// </summary>
+        protected virtual bool ConfinePositionalInput => false;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/CompositeDrawable.cs]: why not `Masking` as default

File: osu.Framework/Graphics/Containers/ScrollContainer.cs [hunk 2]
```diff
@@ -136,6 +136,8 @@ public bool ScrollbarOverlapsContent
 
         protected override Container<T> Content => content;
 
+        protected override bool ConfinePositionalInput => true;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/ScrollContainer.cs]: why is this required?

Comment (nyquillerium) [reply osu.Framework/Graphics/Containers/ScrollContainer.cs]: should be at `BeatmapCarousel` level instead

File: osu.Framework/Graphics/Containers/CompositeDrawable.cs [hunk 3]
```diff
@@ -1221,12 +1221,18 @@ internal override bool BuildNonPositionalInputQueue(List<Drawable> queue, bool a
             return true;
         }
 
+        /// <summary>
+        /// Whether or not this Drawable should block its children from receiving input outside of its
+        /// draw rectangle regardless of whether or not they are masked away.
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/CompositeDrawable.cs]: forgot something

**Comment by smoogipoo** [comment]
I suggest the following structure, since `ConfinePositionalInput` is a bit of a weird name + "draw rectangle" is not what's used.

```csharp
/// <summary>
/// Determines whether the subtree of this <see cref="CompositeDrawable"/> should receive positional input when the mouse is at the given screen-space position.
/// </summary>
/// <remarks>
/// By default, the subtree of this <see cref="CompositeDrawable"/> always receives input when masking is turned off, and only receives input if this
/// <see cref="CompositeDrawable"/> also receives input when masking is turned on.
/// </remarks>
/// <param name="screenSpacePos">The screen-space position where input could be received.</param>
/// <returns>True if the subtree should receive input at the given screen-space position.</returns>
protected virtual bool ReceiveSubTreePositionalInputAt(Vector2 screenSpacePos) => !Masking || ReceivePositionalInputAt(screenSpacePos);

internal override bool BuildPositionalInputQueue(Vector2 screenSpacePos, List<Drawable> queue)
{
    ...

    if (!ReceiveSubTreePositionalInputAt(screenSpacePos))
        return false;

    ...

    return true;
}
```
